front_end/test/downloads.js

```diff
@@ -0,0 +1,187 @@
+/**
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as assert from 'assert';
+import { Downloads } from "../src/backend/downloads.js";
+import * as sinon from 'sinon';
+import axios from 'axios';
+import { describe, it } from 'mocha';
+
+async function fakeAxiosGet(URL) {
+    // Use a specific "wrong URL" to mimic a GET request failing. Return -1 as specified in downloads.js
+    const wrongURL = `https://storage.googleapis.com/wrongBucket/wrongFile`;
+    if (URL == wrongURL) {
+        return -1;
+    }
+
+    // If the URL does not match https://storage.googleapis.com/${bucketName}/${fileName}, throw an error.
+    const regexURL = 'https:\/\/storage\.googleapis\.com\/([a-zA-Z]+(-[a-zA-Z]+)+)[0-9]+\/[0-9]+([a-zA-Z]+(\.[a-zA-Z]+)+)';
+    if (!URL.match(regexURL)) {
+        throw new Error('URL is not well defined');
+    }
+}
+
+function fakeDateNow() {
+    return 1;
+}
+
+describe('downloads', () => {
+    let downloads;
+
+    sinon.stub(Date, "now").callsFake(fakeDateNow);
+    sinon.stub(axios, "get").callsFake(fakeAxiosGet);
+
+    beforeEach(() => {
+        downloads = new Downloads();
+    });
+
+    describe('downloadFile', () => {
+        it('should return -1 on failure of GET request', async () => {
+            const URL = 'poorly_formed_URL';
+            const result = await downloads.downloadFile(URL);
+            assert.equal(result, -1);
+        });
+
+        it('should return 0 on success', async () => {
+            const URL = 'https://storage.googleapis.com/gcsrbpa-us-west1/2mib.txt';
+            const result = await downloads.downloadFile(URL);
+            assert.equal(result, 0);
+        });
+    });
+
+    describe('timeDownload', () => {
+        it('should throw Error if bucketName is invalid', async () => {
+            const fileName = "2mib.txt";
+            const bucketName = "random_bucket_name";
+
+            assert.rejects(downloads.timeDownload(fileName, bucketName), "Invalid Bucket Name");
+        });
+
+        it('should throw Error if fileName is invalid', async () => {
+            const fileName = "random_file_name";
+            const bucketName = "us-west1";
+
+            assert.rejects(downloads.timeDownload(fileName, bucketName), "Invalid File Name");
+        });
+
+        it('should return -0.001 if GET request fails', async () => {
```

Comment:
why would it return -.001?

Reply:
If `downloadFile` fails, it returns `-1`, which is then divided by 1000 to convert from milliseconds to seconds.